ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/StateMachine.hs

```diff
@@ -0,0 +1,810 @@
+               lif
+               cfg
+               (MempoolCapacityBytesOverride $ MempoolCapacityBytes 400)
+               nullTracer
+               txInBlockSize
+  pure (SUT mempool t)
+
+semantics ::
+     (MonadSTM m, LedgerSupportsMempool blk) =>
+     Command blk Concrete
+  -> StrictTVar m (SUT m blk)
+  -> m (Response blk Concrete)
+semantics cmd r = do
+  SUT m t <- atomically $ readTVar r
+  case cmd of
+    Action (NewTxs txs) -> do
+      atomically $ do
+       Tips ti (_, reachable) <- readTVar t
+       -- The mempool tip is not reachable so it will fail. Already change it
+       when (not reachable) $ writeTVar t (Tips ti (ti, True))
+      void $ tryAddTxs m DoNotIntervene txs
+      pure Void
+
+    Action SyncLedger   -> do
+     void $ syncWithLedger m
+     atomically $ do
+       Tips ti _ <- readTVar t
+       -- We just synced with the ledger
+       writeTVar t (Tips ti (ti, True))
+     pure Void
+
+    Action GetSnapshot -> do
+      snap <- atomically (getSnapshot m)
+      pure
+        . GotSnapshot
+        $ Snap [ (txForgetValidated tx, tk)
+               | (tx, tk) <- snapshotTxs snap
+               ]
+
+    Event (ChangeLedger l' becomesUnreachable) -> atomically $ do
+       Tips _ (ot, reachable) <- readTVar t
+       let ot' = if reachable
+                 then (ot, not becomesUnreachable)
+                 else (ot, reachable)
+       writeTVar t (Tips l' ot')
+       pure Void
+
+{-------------------------------------------------------------------------------
+  Conditions
+-------------------------------------------------------------------------------}
+
+noPreconditions :: Model blk Symbolic -> Command blk Symbolic -> Logic
+noPreconditions = const $ const Top
+
+postconditions ::
+  ( LedgerSupportsMempool blk
+  , Eq (GenTx blk)
+  )
+  => Model    blk Concrete
+  -> Command  blk Concrete
+  -> Response blk Concrete
+  -> Logic
+postconditions model (Action GetSnapshot) (GotSnapshot (Snap txs)) =
+  modelTxs model .== txs
+postconditions _ _ _ = Top
+
+shrinker :: Model blk Symbolic
+         -> Command blk Symbolic
+         -> [Command blk Symbolic]
+shrinker _ (Action (NewTxs txs)) =
+  Action . NewTxs <$> shrinkList shrinkNothing txs
+shrinker _ _ = []
+
+{-------------------------------------------------------------------------------
+  State Machine
+-------------------------------------------------------------------------------}
+
+sm ::
+  ( HasTxId (GenTx blk)
+  , LedgerSupportsMempool blk
+  , Eq (GenTx blk)
+  , Arbitrary (LedgerState blk ValuesMK)
+  , ToExpr (LedgerState blk ValuesMK)
+  , ToExpr (GenTx blk)
+  , Eq (TickedLedgerState blk ValuesMK)
+  , LedgerSupportsProtocol blk
+  , UnTick blk
+  , IOLike m
+  , NoThunks (Mempool m blk)
+  )
+  => LedgerConfig blk
+  -> LedgerState blk ValuesMK
+  -> (Int -> LedgerState blk ValuesMK -> Gen [GenTx blk])
+  -> m (StateMachine (Model blk) (Command blk) m (Response blk))
+sm cfg initialState gTxs = do
```

Comment:
Do we want to generate the initial state, or always start with an empty one. The big disadvantage we have by generating and passing the initial state is that we QSM won't shrink it. If an error is triggered by a certain initial state, perhaps we can rely on QC to generate a "set ledger state" action that will uncover the error.